Download seq torda

vtorda opened the pull request:
I modified the 1_Download_seqs.R script so it is possible to prefilter the sequences before downloading it by minimum and maximum sequence length and by a "Tech" field which could be useful to detect wgs, est, or other non-barcoding molecular data. There can be problems with the downloading (e.g., time out) that can be mitigated by setting up the API key and adjusting the chunk size (the number of sequences that are downloaded in one batch). Too small chunk size could make the fetching unreliable, skipping some target sequences, so a chunk size between 150-200 is recommended. Probably not the chunk size but the number of iterations affect the performance of the entrez_post function.

vtorda committed:
Create 1_Download_seqs.R

Copy the new script

vtorda committed:
an extra script was added to fetch more metadata

This script should be merged with the download script

vtorda committed:
Download seq script is updated

I modified the script to be able to prefilter the sequence data before downloading it. There could be fetching problems, that can be mitigated by setting the chunk argument.

vtorda committed:
Delete 4_Download_meta_extra.R

This script was built in the download script

Rowena-h commented on 1_Download_seqs.R:
```diff
@@ -0,0 +1,146 @@
+      cat("You couldn't download all the sequences, try to increase the chunk size")
+    }
+
+    write.table(x = df2, file = paste0(path_to_output_dir, search_term, ".metadata.tsv"),
```
Currently this throws an error if not all sequences were downloaded (as `df2` never gets created in that scenario). You could either put `write.table` within the ifelse above, or figure out how to make an intermediate df2 with not all downloaded sequences to write to file in that scenario (although maybe that's not so useful anyway?)

vtorda replied:
That is what I was afraid of.. the entrez_post function could have a bug, or its performance depends on the network. I don't know... how many sequences did you aim to download? Have you tried to increase the value of the chunk argument?

Rowena-h replied:
Yeah it's quite variable, sometimes I get them all fine and sometimes some are missing. To be honest it seems to be independent of chunk size. For now I think the most conservative thing is to put `write.table` inside the if statement, and try and come up with a solution for matching the metadata to the number of sequences *actually downloaded* later - from a quick look something like extracting the fasta headers from the `rentrez::entrez_fetch` string output during the download for loop, and then matching to the captions to filter `meta_keep`

vtorda replied:
Ok, then, this is not a viable direction. I have already checked the rentrez GitHub and others have similar problems.  The best would be to avoid the post function for which we should be able to edit the web history that I tried previously, but I had the same problem like this: https://github.com/ropensci/rentrez/issues/144

Unfortunately, if the metadata (all_info2) does not match with the meta_keep, then it has already missed the sequences from the fasta files too. So in my opinion what we can do is:

1.  Download all sequences and filter the data after it: we can avoid the  entrez_post function, which I think is the main problem, and Emily and Brigid had a similar problem with it. As soon as I bypassed this function, I got consistent results (see the previous version of the script)
2. Emily wrote a for loop to download sequences multiple times: I think it is not viable for big data
3. If I understand well, your suggestion is to check the IDs on the fly, so during the for loops. Actually, this gave me the idea that I could just check the missing IDs using the metadata tables and run a while loop with only the few sequence IDs that are skipped for some reason until all the desired IDs are in the extended meta table. It is similar to what Emily did, but not repeating everything, just those that we are missing!
4. If the IDs could be appended to a web history object as mentioned above, the problem would be solved

I will try to go for option 3 in the following days because even with a while loop, it would be more efficient than downloading everything and then filtering (in some cases, there could be thousands of wgs data), and probably the first attempt will miss only just a few sequences. What do you think?

vtorda committed:
Updating the Download_seqs function

It turned out that NAs were introduced at the lapply(summary_result, function(x) x[fields]) step if not all fields were presented in a record. referring to fields by names rather then  selecting the first 27 fields by indices could help the problem

vtorda commented:
I could resolve the previous problem. Yet, I encountered another bug regarding one of the functions that call a curl, which occasionally gives an error: "in curl::curl_fetch_memory(url, handle = handle) : Transferred a partial file [eutils.ncbi.nlm.nih.gov]: transfer closed with outstanding read data remaining" I wonder if it is a machine and system level problem. Nevertheless, if such an error occurs, we must keep repeating until it does the job without error.  Could you check if the function works well now?

Rowena-h commented:
> I could resolve the previous problem. Yet, I encountered another bug regarding one of the functions that call a curl, which occasionally gives an error: "in curl::curl_fetch_memory(url, handle = handle) : Transferred a partial file [eutils.ncbi.nlm.nih.gov]: transfer closed with outstanding read data remaining" I wonder if it is a machine and system level problem. Nevertheless, if such an error occurs, we must keep repeating until it does the job without error. Could you check if the function works well now?

Yes I'm also sporadically getting the curl error, both at the `entrez_link` and `entrez_fetch` steps, but as you say I can't see a workaround other than repeatedly submitting! Otherwise I can generate the same Otidea test files.
